cpp/src/reductions/reduction_dispatcher.cuh

```diff
@@ -0,0 +1,151 @@
+    cudf::reductions::ReductionSumOfSquares op)
+{
+    auto it_raw = cudf::make_iterator<has_nulls, T_in, T_out>(input, identity);
+    return thrust::make_transform_iterator(it_raw,
+        cudf::transformer_squared<T_out>{});
+}
+
+// Reduction for 'sum', 'product', 'min', 'max', 'sum of squares'
+// which directly compute the reduction by a single step reduction call
+template<typename T_in, typename T_out, typename Op, bool has_nulls>
+void ReduceOp(const gdf_column *input,
+                   gdf_scalar* scalar, cudaStream_t stream)
+{
+    T_out identity = Op::Op::template identity<T_out>();
+
+    // allocate temporary memory for the result
+    void *result = NULL;
+    RMM_TRY(RMM_ALLOC(&result, sizeof(T_out), stream));
+
+    // initialize output by identity value
+    CUDA_TRY(cudaMemcpyAsync(result, &identity,
+            sizeof(T_out), cudaMemcpyHostToDevice, stream));
+    CHECK_STREAM(stream);
+
+    // reduction by iterator
+    auto it = make_iterator<has_nulls, T_in, T_out>(*input, identity, Op{});
+    reduction_op(static_cast<T_out*>(result), it, input->size, identity,
+        typename Op::Op{}, stream);
+
+    // read back the result to host memory
+    // TODO: asynchronous copy
+    CUDA_TRY(cudaMemcpy(&scalar->data, result,
+            sizeof(T_out), cudaMemcpyDeviceToHost));
+
+    // cleanup temporary memory
+    RMM_TRY(RMM_FREE(result, stream));
+
+    // set scalar is valid
+    scalar->is_valid = true;
+};
+
+
+template <typename T_in, typename Op>
+struct ReduceOutputDispatcher {
+private:
+    template <typename T_out>
+    static constexpr bool is_supported_v()
+    {
+        // for single step reductions,
+        // the available combination of input and output dtypes are
+        //  - same dtypes (including cudf::wrappers)
+        //  - any arithmetic dtype to any arithmetic dtype
+        //  - cudf::bool8 to/from any arithmetic dtype
+        return  std::is_convertible<T_in, T_out>::value ||
+        ( std::is_arithmetic<T_in >::value && std::is_same<T_out, cudf::bool8>::value ) ||
+        ( std::is_arithmetic<T_out>::value && std::is_same<T_in , cudf::bool8>::value );
+    }
+
+public:
+    template <typename T_out, typename std::enable_if<
+        is_supported_v<T_out>() >::type* = nullptr>
+    void operator()(const gdf_column *col,
+                         gdf_scalar* scalar, cudaStream_t stream)
+    {
+        if( col->valid == nullptr ){
+            ReduceOp<T_in, T_out, Op, false>(col, scalar, stream);
+        }else{
+            ReduceOp<T_in, T_out, Op, true >(col, scalar, stream);
+        }
+    }
+
+    template <typename T_out, typename std::enable_if<
+        not is_supported_v<T_out>() >::type* = nullptr >
+    void operator()(const gdf_column *col,
+                         gdf_scalar* scalar, cudaStream_t stream)
+    {
+        CUDF_FAIL("input data type is not convertible to output data type");
+    }
+};
+
+template <typename Op>
+struct ReduceDispatcher {
```

Comment:
Document that this functor is used to dispatch the input element type. Perhaps rename to `ElementTypeDispatcher`.

Reply:
`element_type_dispatcher`

Reply:
`simple_reduction_element_type_dispatcher`